cmd/oidc-example-rp/server.go

```diff
@@ -0,0 +1,150 @@
+		<h1>Start auth flow</h1>
+		<form action="/start" method="POST">
+    		<input type="submit" value="Submit">
+		</form>
+	</body>
+</html>`
+
+var homeTmpl = template.Must(template.New("loginPage").Parse(homePage))
+
+func (s *server) home(w http.ResponseWriter, req *http.Request) {
+	tmplData := map[string]interface{}{}
+
+	if err := homeTmpl.Execute(w, tmplData); err != nil {
+		http.Error(w, fmt.Sprintf("failed to render template: %v", err), http.StatusInternalServerError)
+		return
+	}
+}
+
+// start the actual flow. this builds up the request and sends the user on
+func (s *server) start(w http.ResponseWriter, req *http.Request) {
+	if err := req.ParseForm(); err != nil {
+		http.Error(w, fmt.Sprintf("failed to parse form: %v", err), http.StatusInternalServerError)
+		return
+	}
+
+	// track a random state var to prevent CSRF
+	state := mustRandStr(16)
+	sc := &http.Cookie{
+		Name:   stateCookie,
+		Value:  state,
+		MaxAge: 60,
+	}
+	http.SetCookie(w, sc)
+
+	http.Redirect(w, req, s.oa2cfg.AuthCodeURL(state), http.StatusSeeOther)
+}
+
+const callbackPage = `<!DOCTYPE html>
+<html>
+	<head>
+		<meta charset="UTF-8">
+		<title>LOG IN</title>
+	</head>
+	<body>
+		<div>access_token: {{ .access_token }}</div>
+		<div>id_token: {{ .id_token }}</div>
+	</body>
+</html>`
+
+var callbackTmpl = template.Must(template.New("loginPage").Parse(callbackPage))
+
+func (s *server) callback(w http.ResponseWriter, req *http.Request) {
+	if err := req.ParseForm(); err != nil {
+		http.Error(w, fmt.Sprintf("failed to parse form: %v", err), http.StatusInternalServerError)
+		return
+	}
+
+	statec, err := req.Cookie(stateCookie)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("failed to get state cookie: %v", err), http.StatusInternalServerError)
+		return
+	}
+
+	if errMsg := req.FormValue("error"); errMsg != "" {
+		http.Error(w, fmt.Sprintf("error returned to callback %s: %s", errMsg, req.FormValue("error_description")), http.StatusInternalServerError)
+		return
+	}
+
+	code := req.FormValue("code")
+	if code == "" {
+		http.Error(w, "no code in callback response", http.StatusBadRequest)
+		return
+	}
+
+	gotState := req.FormValue("state")
+	if gotState == "" || gotState != statec.Value {
+		http.Error(w, fmt.Sprintf("returned state %q doesn't match request state %q", gotState, statec.Value), http.StatusBadRequest)
+		return
+	}
+
+	oa2Tok, err := s.oa2cfg.Exchange(req.Context(), code)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("error exchanging code %q for token: %v", code, err), http.StatusInternalServerError)
+	}
+
+	rawIDToken, ok := oa2Tok.Extra("id_token").(string)
+	if !ok {
+		http.Error(w, "no id_token included in response", http.StatusBadRequest)
+	}
+
+	tmplData := map[string]interface{}{
+		"access_token": oa2Tok.AccessToken,
+		"id_token":     rawIDToken,
+	}
+
+	if err := callbackTmpl.Execute(w, tmplData); err != nil {
+		http.Error(w, fmt.Sprintf("failed to render template: %v", err), http.StatusInternalServerError)
+		return
+	}
```

Comment:
This is only about the 100th time we've written this same code, heh. Looks good though.